File: pack/.claude/skills/devrites-lib/scripts/conventions.py
#!/usr/bin/env python3
"""DevRites conventions ledger — store + deterministic corroboration scorer.

The ledger is a project-scoped, LOCAL record (`.devrites/conventions.md`, gitignored)
of conventions a *sealed slice proved* about this codebase: build/test commands, naming /
layering / error-model idioms, recurring gotchas, where things live. `/rite-seal` promotes
to it on a GO; orient reads it so the wright stops re-deriving the same idioms every slice.

Two things live here (issue B1 modules M1 + M2):

  M1 — corroboration scorer (`band`): a pure, deterministic function from corroboration
       and contradiction counts to a 0.3–0.9 confidence band. The number is EARNED by how
       many independent sealed slices proved the convention — never guessed by a model.

  M2 — ledger store (`read` / `promote` / `contradict`): read/update `.devrites/conventions.md`,
       a human-readable Markdown file with per-entry provenance.

Confidence never raises a convention's *authority*: a high-band entry is still untrusted
data, and a fresh observation of the live code always overrides it (enforced at orient,
issue B2 — not here).

Usage:
  conventions.py band --corroborations C --contradictions K
  conventions.py read [--root DIR]
  conventions.py promote   --key K --statement S --kind KIND --slug SLUG --evidence E [--date ISO] [--root DIR]
  conventions.py contradict --key K --slug SLUG --evidence E [--date ISO] [--root DIR]

Exit codes: 0 ok; 2 usage error; 3 not-found (contradict on an unknown key).
"""
import argparse
import datetime
import os
import sys

LEDGER_RELPATH = os.path.join(".devrites", "conventions.md")

HEADER = """# DevRites conventions ledger

Conventions proven by sealed slices on this project. **Local, gitignored, advisory.**
The confidence band is earned: it reflects how many independent sealed slices corroborated
the convention — it is not a model's guess. A high band is still untrusted data; a fresh
observation of the live code always wins over a stale entry here.
"""


# --- M1: deterministic corroboration scorer -------------------------------

def band(corroborations, contradictions):
    """Pure, total, deterministic. Return a 0.3–0.9 band, or None if retired.

    First proof seeds mid-band; each further independent corroboration raises it
    toward the 0.9 cap; each contradiction lowers it; net non-positive retires it.
    """
    c = int(corroborations)
    k = int(contradictions)
    if c - k <= 0:
        return None  # retired — contradictions caught up with the evidence
    raw = 0.4 + 0.1 * c - 0.2 * k
    return max(0.3, min(0.9, round(raw, 1)))


# --- M2: ledger store -----------------------------------------------------

def _ledger_path(root):
    return os.path.join(root, LEDGER_RELPATH)


def _parse(text):
    """Parse the ledger into {key: entry}. entry = dict with fields + proofs list."""
    entries = {}
    key = None
    for line in text.splitlines():
        if line.startswith("## "):
            key = line[3:].strip()
            entries[key] = {
                "statement": "", "kind": "", "corroborations": 0,
                "contradictions": 0, "status": "active", "proofs": [],
            }
            continue
        if key is None:
            continue
        s = line.strip()
        if s.startswith("- proof:"):
            entries[key]["proofs"].append(s[len("- proof:"):].strip())
        elif s.startswith("- ") and ":" in s:
            fld, val = s[2:].split(":", 1)
            fld, val = fld.strip(), val.strip()
            if fld in ("corroborations", "contradictions"):
                try:
                    entries[key][fld] = int(val)
                except ValueError:
                    pass
            elif fld in ("statement", "kind", "status"):
                entries[key][fld] = val
    return entries


def _serialize(entries):
    out = [HEADER]
    for key in sorted(entries):
        e = entries[key]
        b = band(e["corroborations"], e["contradictions"])
        status = "retired" if b is None else "active"
        e["status"] = status
        band_str = "retired" if b is None else "%.2f" % b
        out.append("## %s" % key)
        out.append("- statement: %s" % e["statement"])
        out.append("- kind: %s" % e["kind"])
        out.append("- band: %s" % band_str)
        out.append("- corroborations: %d" % e["corroborations"])
        out.append("- contradictions: %d" % e["contradictions"])
        out.append("- status: %s" % status)
        for p in e["proofs"]:
            out.append("- proof: %s" % p)
        out.append("")
    return "\n".join(out).rstrip() + "\n"


def _load(root):
    path = _ledger_path(root)
    if not os.path.isfile(path):
        return {}
    with open(path, "r", encoding="utf-8") as fh:
        return _parse(fh.read())


def _save(root, entries):
    path = _ledger_path(root)
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w", encoding="utf-8") as fh:
        fh.write(_serialize(entries))


def _today(date):
    return date or datetime.date.today().isoformat()


def cmd_read(root):
    path = _ledger_path(root)
    if not os.path.isfile(path):
        return 0
    with open(path, "r", encoding="utf-8") as fh:
        sys.stdout.write(fh.read())
    return 0


def cmd_orient(root, min_band):
    """Print a compact digest of ACTIVE conventions for the wright's ORIENT step.

    These are priors, not law: high-band entries are followed unless the slice contract
    says otherwise, and a fresh observation of the live code always overrides a stale one.
    Retired entries are omitted. Empty output (exit 0) means no priors to apply.
    """
    entries = _load(root)
    rows = []
    for key, e in entries.items():
        b = band(e["corroborations"], e["contradictions"])
        if b is None or b < min_band:
            continue
        rows.append((b, key, e))
    if not rows:
        return 0
    rows.sort(key=lambda r: (-r[0], r[1]))
    print("# Conventions ledger — proven priors (live code always overrides)")
    for b, key, e in rows:
        kind = (" (%s)" % e["kind"]) if e["kind"] else ""
        print("- [%.2f] %s%s: %s" % (b, key, kind, e["statement"]))
    return 0


def cmd_promote(root, key, statement, kind, slug, evidence, date):
    entries = _load(root)
    e = entries.get(key)
    if e is None:
        e = {"statement": statement, "kind": kind, "corroborations": 0,
             "contradictions": 0, "status": "active", "proofs": []}
        entries[key] = e
    else:
        # keep the latest non-empty statement/kind
        if statement:
            e["statement"] = statement
        if kind:
            e["kind"] = kind
    # Corroboration counts DISTINCT sealed slices — re-sealing the same slice is idempotent.
    already = any(p.split(" ", 1)[0] == slug for p in e["proofs"])
    if not already:
        e["corroborations"] += 1
        e["proofs"].append("%s %s — %s" % (slug, _today(date), evidence))
    _save(root, entries)
    b = band(e["corroborations"], e["contradictions"])
    print("promoted '%s' → band %s (%d corroboration(s))"
          % (key, "retired" if b is None else "%.2f" % b, e["corroborations"]))
    return 0


def cmd_contradict(root, key, slug, evidence, date, drift_file):
    entries = _load(root)
    e = entries.get(key)
    if e is None:
        print("no such convention: '%s'" % key, file=sys.stderr)
        return 3
    e["contradictions"] += 1
    e["proofs"].append("%s %s — CONTRADICTED: %s" % (slug, _today(date), evidence))
    _save(root, entries)
    b = band(e["corroborations"], e["contradictions"])
    state = "retired" if b is None else "band %.2f" % b
    when = _today(date)
    # The orchestrator (rite-build) is the single writer of .devrites bookkeeping; it passes
    # --drift-file so the contradiction is logged to the feature's drift.md atomically here.
    if drift_file:
        os.makedirs(os.path.dirname(os.path.abspath(drift_file)), exist_ok=True)
        with open(drift_file, "a", encoding="utf-8") as fh:
            fh.write(
                "\n## convention-drift: %s\n"
                "- slice: %s\n- contradicted_by: %s\n- now: %s\n- at: %s\n"
                % (key, slug, evidence, state, when))
    print("DRIFT: convention '%s' contradicted by %s — now %s" % (key, slug, state))
    return 0


def main(argv):
    p = argparse.ArgumentParser(prog="conventions.py", add_help=True)
    sub = p.add_subparsers(dest="cmd")

    pb = sub.add_parser("band")
    pb.add_argument("--corroborations", type=int, required=True)
    pb.add_argument("--contradictions", type=int, default=0)

    pr = sub.add_parser("read")
    pr.add_argument("--root", default=".")

    po = sub.add_parser("orient")
    po.add_argument("--root", default=".")
    po.add_argument("--min-band", type=float, default=0.0)

    pp = sub.add_parser("promote")
    pp.add_argument("--key", required=True)
    pp.add_argument("--statement", required=True)
    pp.add_argument("--kind", required=True)
    pp.add_argument("--slug", required=True)
    pp.add_argument("--evidence", required=True)
    pp.add_argument("--date", default=None)
    pp.add_argument("--root", default=".")

    pc = sub.add_parser("contradict")
    pc.add_argument("--key", required=True)
    pc.add_argument("--slug", required=True)
    pc.add_argument("--evidence", required=True)
    pc.add_argument("--date", default=None)
    pc.add_argument("--root", default=".")
    pc.add_argument("--drift-file", default=None)

    args = p.parse_args(argv[1:])
    if args.cmd == "band":
        b = band(args.corroborations, args.contradictions)
        print("retired" if b is None else "%.2f" % b)
        return 0
    if args.cmd == "read":
        return cmd_read(args.root)
    if args.cmd == "orient":
        return cmd_orient(args.root, args.min_band)
    if args.cmd == "promote":
        return cmd_promote(args.root, args.key, args.statement, args.kind,
                           args.slug, args.evidence, args.date)
    if args.cmd == "contradict":
        return cmd_contradict(args.root, args.key, args.slug, args.evidence, args.date,
                              args.drift_file)
    p.print_usage(sys.stderr)
    return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: engine/internal/lib/conventions.go
package lib

import (
	"bytes"
	"embed"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"text/template"
	"time"
)

//go:embed templates/conventions_ledger_header.md templates/conventions_orient_header.txt templates/conventions_drift.tmpl
var conventionsTemplateFS embed.FS

var (
	conventionsHeader       = mustReadConventionsTemplate("templates/conventions_ledger_header.md")
	conventionsOrientHeader = mustReadConventionsTemplate("templates/conventions_orient_header.txt")
	conventionsDriftTmpl    = template.Must(template.ParseFS(conventionsTemplateFS, "templates/conventions_drift.tmpl"))
)

type conventionEntry struct {
	Statement      string
	Kind           string
	Status         string
	Proofs         []string
	Corroborations int
	Contradictions int
}

func Conventions(args []string, stdout, stderr io.Writer) int {
	cmd := argAt(args, 0)
	switch cmd {
	case "band":
		c, ok := intFlag(args[1:], "--corroborations", true, stderr)
		if !ok {
			return 2
		}
		k, ok := intFlag(args[1:], "--contradictions", false, stderr)
		if !ok {
			return 2
		}
		if b, active := conventionBand(c, k); active {
			fmt.Fprintf(stdout, "%.2f\n", b)
		} else {
			fmt.Fprintln(stdout, "retired")
		}
		return 0
	case "read":
		root := stringFlag(args[1:], "--root", ".")
		data, err := os.ReadFile(conventionsPath(root))
		if err != nil {
			return 0
		}
		_, _ = stdout.Write(data)
		return 0
	case "orient":
		return conventionsOrient(args[1:], stdout, stderr)
	case "promote":
		return conventionsPromote(args[1:], stdout, stderr)
	case "contradict":
		return conventionsContradict(args[1:], stdout, stderr)
	default:
		fmt.Fprintln(stderr, "usage: devrites-engine conventions <band|read|orient|promote|contradict> [flags]")
		return 2
	}
}

func conventionBand(corroborations, contradictions int) (float64, bool) {
	if corroborations-contradictions <= 0 {
		return 0, false
	}
	raw := 0.4 + 0.1*float64(corroborations) - 0.2*float64(contradictions)
	return math.Max(0.3, math.Min(0.9, math.Round(raw*10)/10)), true
}

func conventionsOrient(args []string, stdout, stderr io.Writer) int {
	root := stringFlag(args, "--root", ".")
	minBand := floatFlag(args, "--min-band", 0)
	entries := loadConventions(root)
	type row struct {
		key string
		b   float64
		e   conventionEntry
	}
	var rows []row
	for key, e := range entries {
		b, active := conventionBand(e.Corroborations, e.Contradictions)
		if !active || b < minBand {
			continue
		}
		rows = append(rows, row{key: key, b: b, e: e})
	}
	if len(rows) == 0 {
		return 0
	}
	sort.Slice(rows, func(i, j int) bool {
		if rows[i].b != rows[j].b {
			return rows[i].b > rows[j].b
		}
		return rows[i].key < rows[j].key
	})
	fmt.Fprintln(stdout, conventionsOrientHeader)
	for _, r := range rows {
		kind := ""
		if r.e.Kind != "" {
			kind = " (" + r.e.Kind + ")"
		}
		fmt.Fprintf(stdout, "- [%.2f] %s%s: %s\n", r.b, r.key, kind, r.e.Statement)
	}
	return 0
}

func conventionsPromote(args []string, stdout, stderr io.Writer) int {
	root := stringFlag(args, "--root", ".")
	key, ok := requiredStringFlag(args, "--key", stderr)
	if !ok {
		return 2
	}
	statement, ok := requiredStringFlag(args, "--statement", stderr)
	if !ok {
		return 2
	}
	kind, ok := requiredStringFlag(args, "--kind", stderr)
	if !ok {
		return 2
	}
	slug, ok := requiredStringFlag(args, "--slug", stderr)
	if !ok {
		return 2
	}
	evidence, ok := requiredStringFlag(args, "--evidence", stderr)
	if !ok {
		return 2
	}
	date := stringFlag(args, "--date", "")
	entries := loadConventions(root)
	e, exists := entries[key]
	if !exists {
		e = conventionEntry{Statement: statement, Kind: kind, Status: "active"}
	} else {
		if statement != "" {
			e.Statement = statement
		}
		if kind != "" {
			e.Kind = kind
		}
	}
	already := false
	for _, proof := range e.Proofs {
		if proofSlug(proof) == slug {
			already = true
			break
		}
	}
	if !already {
		e.Corroborations++
		e.Proofs = append(e.Proofs, fmt.Sprintf("%s %s — %s", slug, conventionDate(date), evidence))
	}
	entries[key] = e
	if err := saveConventions(root, entries); err != nil {
		fmt.Fprintln(stderr, err)
		return 1
	}
	b, active := conventionBand(e.Corroborations, e.Contradictions)
	band := "retired"
	if active {
		band = fmt.Sprintf("%.2f", b)
	}
	fmt.Fprintf(stdout, "promoted '%s' → band %s (%d corroboration(s))\n", key, band, e.Corroborations)
	return 0
}

func conventionsContradict(args []string, stdout, stderr io.Writer) int {
	root := stringFlag(args, "--root", ".")
	key, ok := requiredStringFlag(args, "--key", stderr)
	if !ok {
		return 2
	}
	slug, ok := requiredStringFlag(args, "--slug", stderr)
	if !ok {
		return 2
	}
	evidence, ok := requiredStringFlag(args, "--evidence", stderr)
	if !ok {
		return 2
	}
	date := stringFlag(args, "--date", "")
	driftFile := stringFlag(args, "--drift-file", "")
	entries := loadConventions(root)
	e, exists := entries[key]
	if !exists {
		fmt.Fprintf(stderr, "no such convention: '%s'\n", key)
		return 3
	}
	when := conventionDate(date)
	e.Contradictions++
	e.Proofs = append(e.Proofs, fmt.Sprintf("%s %s — CONTRADICTED: %s", slug, when, evidence))
	entries[key] = e
	if err := saveConventions(root, entries); err != nil {
		fmt.Fprintln(stderr, err)
		return 1
	}
	b, active := conventionBand(e.Corroborations, e.Contradictions)
	state := "retired"
	if active {
		state = fmt.Sprintf("band %.2f", b)
	}
	if driftFile != "" {
		if err := os.MkdirAll(filepath.Dir(absPath(driftFile)), 0o755); err != nil {
			fmt.Fprintln(stderr, err)
			return 1
		}
		f, err := os.OpenFile(driftFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
		if err != nil {
			fmt.Fprintln(stderr, err)
			return 1
		}
		entry, renderErr := renderConventionDrift(key, slug, evidence, state, when)
		if renderErr != nil {
			_ = f.Close()
			fmt.Fprintln(stderr, renderErr)
			return 1
		}
		_, writeErr := f.WriteString(entry)
		closeErr := f.Close()
		if writeErr != nil {
			fmt.Fprintln(stderr, writeErr)
			return 1
		}
		if closeErr != nil {
			fmt.Fprintln(stderr, closeErr)
			return 1
		}
	}
	fmt.Fprintf(stdout, "DRIFT: convention '%s' contradicted by %s — now %s\n", key, slug, state)
	return 0
}

func conventionsPath(root string) string {
	return filepath.Join(root, ".devrites", "conventions.md")
}

func loadConventions(root string) map[string]conventionEntry {
	data, err := os.ReadFile(conventionsPath(root))
	if err != nil {
		return map[string]conventionEntry{}
	}
	return parseConventions(string(data))
}

func parseConventions(text string) map[string]conventionEntry {
	entries := map[string]conventionEntry{}
	key := ""
	for _, line := range strings.Split(text, "\n") {
		if strings.HasPrefix(line, "## ") {
			key = strings.TrimSpace(strings.TrimPrefix(line, "## "))
			entries[key] = conventionEntry{Status: "active"}
			continue
		}
		if key == "" {
			continue
		}
		e := entries[key]
		s := strings.TrimSpace(line)
		switch {
		case strings.HasPrefix(s, "- proof:"):
			e.Proofs = append(e.Proofs, strings.TrimSpace(strings.TrimPrefix(s, "- proof:")))
		case strings.HasPrefix(s, "- ") && strings.Contains(s, ":"):
			field, value, _ := strings.Cut(strings.TrimPrefix(s, "- "), ":")
			field = strings.TrimSpace(field)
			value = strings.TrimSpace(value)
			switch field {
			case "corroborations":
				if n, err := strconv.Atoi(value); err == nil {
					e.Corroborations = n
				}
			case "contradictions":
				if n, err := strconv.Atoi(value); err == nil {
					e.Contradictions = n
				}
			case "statement":
				e.Statement = value
			case "kind":
				e.Kind = value
			case "status":
				e.Status = value
			}
		}
		entries[key] = e
	}
	return entries
}

func saveConventions(root string, entries map[string]conventionEntry) error {
	path := conventionsPath(root)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return writeFileAtomic(path, []byte(serializeConventions(entries)), 0o644)
}

func serializeConventions(entries map[string]conventionEntry) string {
	var keys []string
	for key := range entries {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	var out []string
	out = append(out, conventionsHeader)
	for _, key := range keys {
		e := entries[key]
		b, active := conventionBand(e.Corroborations, e.Contradictions)
		status := "retired"
		band := "retired"
		if active {
			status = "active"
			band = fmt.Sprintf("%.2f", b)
		}
		out = append(out,
			"## "+key,
			"- statement: "+e.Statement,
			"- kind: "+e.Kind,
			"- band: "+band,
			fmt.Sprintf("- corroborations: %d", e.Corroborations),
			fmt.Sprintf("- contradictions: %d", e.Contradictions),
			"- status: "+status,
		)
		for _, proof := range e.Proofs {
			out = append(out, "- proof: "+proof)
		}
		out = append(out, "")
	}
	return strings.TrimRight(strings.Join(out, "\n"), "\n") + "\n"
}

func mustReadConventionsTemplate(path string) string {
	data, err := conventionsTemplateFS.ReadFile(path)
	if err != nil {
		panic(err)
	}
	return strings.TrimRight(string(data), "\n")
}

func renderConventionDrift(key, slug, evidence, state, when string) (string, error) {
	var buf bytes.Buffer
	err := conventionsDriftTmpl.Execute(&buf, struct {
		Key      string
		Slug     string
		Evidence string
		State    string
		When     string
	}{
		Key:      key,
		Slug:     slug,
		Evidence: evidence,
		State:    state,
		When:     when,
	})
	return buf.String(), err
}

func proofSlug(proof string) string {
	slug, _, _ := strings.Cut(proof, " ")
	return slug
}

func conventionDate(date string) string {
	if date != "" {
		return date
	}
	return time.Now().Format("2006-01-02")
}

func stringFlag(args []string, name, fallback string) string {
	value, ok := flagValue(args, name)
	if !ok {
		return fallback
	}
	return value
}

func flagValue(args []string, name string) (string, bool) {
	for i := 0; i < len(args); i++ {
		if args[i] == name && i+1 < len(args) {
			return args[i+1], true
		}
		if strings.HasPrefix(args[i], name+"=") {
			return strings.TrimPrefix(args[i], name+"="), true
		}
	}
	return "", false
}

func requiredStringFlag(args []string, name string, stderr io.Writer) (string, bool) {
	value, ok := flagValue(args, name)
	if !ok {
		fmt.Fprintf(stderr, "missing required flag: %s\n", name)
		return "", false
	}
	return value, true
}

func intFlag(args []string, name string, required bool, stderr io.Writer) (int, bool) {
	value, present := flagValue(args, name)
	if !present {
		if required {
			fmt.Fprintf(stderr, "missing required flag: %s\n", name)
			return 0, false
		}
		return 0, true
	}
	n, err := strconv.Atoi(value)
	if err != nil {
		fmt.Fprintf(stderr, "invalid integer for %s: %s\n", name, value)
		return 0, false
	}
	return n, true
}

func floatFlag(args []string, name string, fallback float64) float64 {
	value := stringFlag(args, name, "")
	if value == "" {
		return fallback
	}
	f, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return fallback
	}
	return f
}

func absPath(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return path
	}
	return abs
}
